lightning/src/events/mod.rs

```diff
@@ -513,6 +513,23 @@ pub enum Event {
 		/// serialized prior to LDK version 0.0.117.
 		sender_intended_total_msat: Option<u64>,
 	},
+	/// Indicates that a peer connection with a node is needed in order to send an [`OnionMessage`].
+	///
+	/// Typically, this happens when a [`MessageRouter`] is unable to find a complete path to a
+	/// [`Destination`]. Once a connection is established, any messages buffered by an
+	/// [`OnionMessageHandler`] may be sent.
+	///
+	/// This event will not be generated for onion message forwards; only for sends including
+	/// replies. Handlers should connect to the node otherwise any buffered messages may be lost.
+	///
+	/// [`OnionMessage`]: msgs::OnionMessage
+	/// [`MessageRouter`]: crate::onion_message::MessageRouter
+	/// [`Destination`]: crate::onion_message::Destination
+	/// [`OnionMessageHandler`]: crate::ln::msgs::OnionMessageHandler
+	ConnectionNeeded {
```

Comment:
It would require adding another parameterization for users to implement (i.e., either a new trait on `OnionMessenger` or one on `DefaultMessageRotuer`), which would then need to bubble up to the `Simple*` aliases.

And since it needs to be async, the trait implementation wouldn't be enough. Users would need to write code to process connection requests in the background from what their trait enqueued. If we wrote some of this logic for them, then we'd need to do it for each language's IO layer. But that wouldn't fit well into any peer management logic that they may have (e.g., LDK Node's `PeerStore`).

Seems much simpler just to surface an event for them so that no additional parameterization and processing code is needed, other than handling one more event variant.

Reply:
Could we pipe the needs-connect nodes through `OffersMessageHandler` / `CustomOMHandler` with a new trait method? Then the `ChannelManager` could generate `ConnectionNeeded` events and we wouldn't have to trouble the background processor. I do see Jeff's argument that the current design aligns with the peer manager doing peer stuff though.

Reply:
Hmm... that wouldn't work for uses of `send_onion_message` or anyone using something other than `ChannelManager` to implement one of those traits.

Reply:
The custom trait implementations would need to support connecting to peers outbound, would it not work in some other way? I'm not convinced it's cleaner than the current approach though :/ just nicer in terms of the `PeerMan`/`BgProcessor`.

Reply:
Not sure that I follow. All custom implementations should work with the current approach of using `OnionMessenger`. Or are you saying a custom implementation of `OnionMessageHandler` (i.e., something other than `OnionMessenger`)? Note that `CustomOnionMessageHandler` and `OffersMessageHandler` are specific to `OnionMessenger`. Any implementations of those work fine with the current approach.

Reply:
Discussed offline. Updated to have events processed directly from any `OnionMessageHandler` by requiring implementations to also implement `EventsProvider`. I was able to do so without adding another type parameter to `BackgroundProcessor` since the handler can be fetched from `PeerManager`.